File: examplencpol.py
# -*- coding: utf-8 -*-
"""
An example that exports to sparse SDPA format for scalable computation. The
description of the example is in the following paper:

Pironio, S.; Navascués, M. & Acín, A. Convergent relaxations of 
polynomial optimization problems with noncommuting variables SIAM Journal
on Optimization, SIAM, 2010, 20, 2157-2180.

Created on Fri May 10 09:45:11 2013

@author: Peter Wittek
"""

from ncutils import generate_ncvariables
from sdprelaxation import get_relaxation

#Number of Hermitian variables    
n_vars = 2
#Order of relaxation
order = 2

#Get Hermitian variables
X = generate_ncvariables(n_vars)

#Define the objective function
obj = X[0] * X[1] + X[1] * X[0]

# Inequality constraints
inequalities = [0] * 1
inequalities[0] = -X[1]**2 + X[1] + 0.5 

# Equality constraints
equalities = []

#Simple monomial substitutions
monomial_substitution = {}
monomial_substitution[X[0]**2] = X[0]

#Obtain SDP relaxation
prob = get_relaxation(X, obj, inequalities, equalities, 
                      monomial_substitution, order)

#Export relaxation to SDPA format
prob.write_to_file('examplenc.dat-s')

File: matlab/examplencpol.m
%EXAMPLE - An example solved with SeDuMi or exported to
% sparse SDPA format for scalable computation. The description 
% of the example is in the following paper:
%
% Pironio, S.; Navascués, M. & Acín, A. Convergent relaxations of 
% polynomial optimization problems with noncommuting variables SIAM Journal
% on Optimization, SIAM, 2010, 20, 2157-2180

% Other m-files required: Yalmip, SeDuMi, ncmoments.m, momentstosdpa.m
% MAT-files required: none
%

% Author: Peter Wittek
% December 2012; Last revision: 31-May-2012

%------------- BEGIN CODE --------------

% Define non-commutative variables
n = 2;
X = ncvar(n,1);

% Cost function
obj = X(1)*X(2)+X(2)*X(1);

% Equality constraints
equalities = X(1)^2-X(1);
% Commutative variant
%equalities = [ X(1)^2-X(1); X(2)*X(1)-X(1)*X(2) ];


% Inequality constraints
inequalities = -X(2)^2+X(2)+1/2;

%Order of relaxation
order = 2;

[Fmoments] = ncmoments(X, inequalities, equalities, order);

%There are two options:

%1) The problem is of small or moderate size. Solve the problem in Matlab.
%   Note that noncommutative variables are experimental, and evaluating the
%   the objective function by double(obj) will not return the correct result.    
yalmip_options = sdpsettings('solver','sedumi','removeequalities ', 1,'relax',1,'saveduals',0);
solvesdp(Fmoments,obj,yalmip_options);

%2) The problem is large. Export it to SDPA sparse format and solve the
%   problem with SDPARA.
momentstosdpa('example.dat-s', Fmoments, obj);

%------------- END OF CODE --------------
